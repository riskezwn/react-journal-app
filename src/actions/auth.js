import {
    createUserWithEmailAndPassword,
    signInWithEmailAndPassword,
    signInWithPopup,
    signOut,
    updateProfile,
} from "firebase/auth";
import Swal from "sweetalert2";
import { auth, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { notesLogout } from "./notes";
import { finishLoading, startLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());
        signInWithEmailAndPassword(auth, email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());
            })
            .catch((e) => {
                Swal.fire("Something go wrong :(", "User not found", "error");
                dispatch(finishLoading());
            });
    };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        dispatch(startLoading());
        createUserWithEmailAndPassword(auth, email, password)
            .then(async ({ user }) => {
                await updateProfile(user, { displayName: name });
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());
            })
            .catch((e) => {
                Swal.fire(
                    "Something go wrong :(",
                    "Registration could not be completed",
                    "error"
                );
                dispatch(finishLoading());
            });
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
            dispatch(login(user.uid, user.displayName));
        });
    };
};

export const startLogout = () => {
    return async (dispatch) => {
        await signOut(auth);
        dispatch(logout());
        dispatch(notesLogout());
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName,
    },
});

export const logout = () => ({
    type: types.logout,
});
