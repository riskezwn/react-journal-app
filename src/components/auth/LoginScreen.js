import React from "react";
import validator from "validator";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { startGoogleLogin, startLoginEmailPassword } from "../../actions/auth";
import { useForm } from "../../hooks/useForm";
import { removeError, setError } from "../../actions/ui";

export const LoginScreen = () => {
    const dispatch = useDispatch();
    const { msgError, loading } = useSelector((state) => state.ui);

    const [formValues, handleInputChange] = useForm({
        email: "riskezwn@gmail.com",
        password: "123456",
    });
    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startLoginEmailPassword(email, password));
        }
    };

    const isFormValid = () => {
        if (!validator.isEmail(email)) {
            dispatch(setError("Email is not valid"));
            return false;
        }
        dispatch(removeError());
        return true;
    };

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    };

    return (
        <div>
            <h2 className="auth__title">Login</h2>
            <form
                onSubmit={handleLogin}
                className="animate__animated animate__fadeIn animate__fast"
            >
                {msgError && (
                    <div className="auth__alert-error">{msgError}</div>
                )}
                <input
                    type="email"
                    placeholder="john@doe.com"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={handleInputChange}
                />
                <input
                    type="password"
                    placeholder="********"
                    name="password"
                    className="auth__input"
                    autoComplete="off"
                    value={password}
                    onChange={handleInputChange}
                />
                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled={loading}
                >
                    Login
                </button>
                <hr />
                <div className="auth__social-networks">
                    <p>Login with social networks:</p>
                    <div className="google-btn" onClick={handleGoogleLogin}>
                        <div className="google-icon-wrapper">
                            <img
                                className="google-icon"
                                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                                alt="google button"
                            />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with Google</b>
                        </p>
                    </div>
                </div>
                <Link to="/auth/register" className="link">
                    Create new account
                </Link>
            </form>
        </div>
    );
};
