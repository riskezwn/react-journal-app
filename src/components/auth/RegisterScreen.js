import React from "react";
import validator from "validator";
import { Link } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { useDispatch, useSelector } from "react-redux";
import { removeError, setError } from "../../actions/ui";
import { startRegisterWithEmailPasswordName } from "../../actions/auth";

export const RegisterScreen = () => {
    const dispatch = useDispatch();
    const { msgError, loading } = useSelector((state) => state.ui);

    const [formValues, handleInputChange] = useForm({
        name: "riskezwn",
        email: "riskezwn@gmail.com",
        password: "123456",
        password2: "123456",
    });
    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();

        if (isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name));
        }
    };

    const isFormValid = () => {
        if (name.trim().length < 1) {
            dispatch(setError("Name is required"));
            return false;
        } else if (!validator.isEmail(email)) {
            dispatch(setError("Email is not valid"));
            return false;
        } else if (password.length < 5) {
            dispatch(setError("Password must be 6 characters long"));
            return false;
        } else if (password !== password2) {
            dispatch(setError("Passwords must match"));
            return false;
        }
        dispatch(removeError());
        return true;
    };

    return (
        <div>
            <h2 className="auth__title">Register</h2>
            <form onSubmit={handleRegister}>
                {msgError && (
                    <div className="auth__alert-error">{msgError}</div>
                )}
                <input
                    type="text"
                    placeholder="John"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={name}
                />
                <input
                    type="email"
                    placeholder="john@doe.com"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={email}
                />
                <input
                    type="password"
                    placeholder="********"
                    name="password"
                    className="auth__input"
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={password}
                />
                <input
                    type="password"
                    placeholder="Confirm your password"
                    name="password2"
                    className="auth__input"
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={password2}
                />
                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                    disabled={loading}
                >
                    Register
                </button>

                <Link to="/auth/login" className="link">
                    Already registered?
                </Link>
            </form>
        </div>
    );
};
