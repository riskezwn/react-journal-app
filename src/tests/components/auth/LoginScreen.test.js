/**
 * @jest-environment jsdom
 */

import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { MemoryRouter } from "react-router-dom";
import { LoginScreen } from "../../../components/auth/LoginScreen";
import "../../../setupTest";
import {
  startGoogleLogin,
  startLoginEmailPassword,
} from "../../../actions/auth";

jest.mock("../../../actions/auth", () => ({
  startGoogleLogin: jest.fn(),
  startLoginEmailPassword: jest.fn(),
}));
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null,
  },
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <LoginScreen />
    </MemoryRouter>
  </Provider>
);
describe("test in <LoginScreen/>", () => {
  beforeEach(() => {
    store = mockStore(initState);
    jest.clearAllMocks();
  });

  test("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should start starLoginScreen with Google button", () => {
    wrapper.find(".google-btn").prop("onClick")();

    expect(startGoogleLogin).toHaveBeenCalled();
  });

  test("should start startLogin with their respective arguments", () => {
    wrapper.find("form").prop("onSubmit")({
      preventDefault() {},
    });

    // expect(startLoginEmailPassword).toHaveBeenCalledWith("", "");
  });
});
