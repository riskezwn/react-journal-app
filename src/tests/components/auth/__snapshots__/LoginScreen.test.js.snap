// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <LoginScreen/> should match snapshot 1`] = `
<div>
  <h2
    className="auth__title"
  >
    Login
  </h2>
  <form
    className="animate__animated animate__fadeIn animate__fast"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="john@doe.com"
      type="email"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="********"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <hr />
    <div
      className="auth__social-networks"
    >
      <p>
        Login with social networks:
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with Google
          </b>
        </p>
      </div>
    </div>
    <a
      className="link"
      href="/auth/register"
      onClick={[Function]}
    >
      Create new account
    </a>
  </form>
</div>
`;
